`merge.py` fails tests without optional dependencies installed

I just ran `make check` on a completely fresh clone of current `master` (`7db90434187`), and it reported 10 errors:

```
======================================================================
ERROR: test_merge_with_conflicts (tests.test_cli_merge.CLIMergeTests.test_merge_with_conflicts)
Test CLI merge with conflicts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_cli_merge.py", line 112, in test_merge_with_conflicts
    retcode = main(["merge", "feature"])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/cli.py", line 3966, in main
    return cmd_kls().run(cmd_args)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/cli.py", line 2297, in run
    merge_commit_id, conflicts = porcelain.merge(
                                 ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4519, in merge
    result = _do_merge(
             ^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4432, in _do_merge
    merged_tree, conflicts = three_way_merge(
                             ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 503, in three_way_merge
    return merger.merge_trees(base_tree, ours_tree, theirs_tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_blobs_driver_not_found (tests.test_merge_drivers.MergeBlobsWithDriversTests.test_merge_blobs_driver_not_found)
Test merge_blobs when specified driver is not found.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_merge_drivers.py", line 318, in test_merge_blobs_driver_not_found
    result, has_conflicts = merge_blobs(
                            ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_blobs_with_text_driver (tests.test_merge_drivers.MergeBlobsWithDriversTests.test_merge_blobs_with_text_driver)
Test merge_blobs with 'text' merge driver (default).
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_merge_drivers.py", line 241, in test_merge_blobs_with_text_driver
    result, has_conflicts = merge_blobs(
                            ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_blobs_without_driver (tests.test_merge_drivers.MergeBlobsWithDriversTests.test_merge_blobs_without_driver)
Test merge_blobs without any merge driver.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_merge_drivers.py", line 224, in test_merge_blobs_without_driver
    result, has_conflicts = merge_blobs(base, ours, theirs)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_cherry_pick_abort (tests.test_porcelain_cherry_pick.PorcelainCherryPickTests.test_cherry_pick_abort)
Test aborting a cherry-pick.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_porcelain_cherry_pick.py", line 195, in test_cherry_pick_abort
    porcelain.cherry_pick(tmpdir, feature_commit)
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4723, in cherry_pick
    merged_tree, conflicts = three_way_merge(
                             ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 503, in three_way_merge
    return merger.merge_trees(base_tree, ours_tree, theirs_tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_cherry_pick_conflict (tests.test_porcelain_cherry_pick.PorcelainCherryPickTests.test_cherry_pick_conflict)
Test cherry-pick with conflicts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_porcelain_cherry_pick.py", line 138, in test_cherry_pick_conflict
    porcelain.cherry_pick(tmpdir, feature_commit)
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4723, in cherry_pick
    merged_tree, conflicts = three_way_merge(
                             ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 503, in three_way_merge
    return merger.merge_trees(base_tree, ours_tree, theirs_tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_with_conflicts (tests.test_porcelain_merge.PorcelainMergeTests.test_merge_with_conflicts)
Test merge with conflicts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_porcelain_merge.py", line 196, in test_merge_with_conflicts
    merge_commit, conflicts = porcelain.merge(tmpdir, "feature")
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4519, in merge
    result = _do_merge(
             ^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4432, in _do_merge
    merged_tree, conflicts = three_way_merge(
                             ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 503, in three_way_merge
    return merger.merge_trees(base_tree, ours_tree, theirs_tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_tree_with_conflicts (tests.test_porcelain_merge.PorcelainMergeTreeTests.test_merge_tree_with_conflicts)
Test merge_tree with conflicts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_porcelain_merge.py", line 359, in test_merge_tree_with_conflicts
    merged_tree_id, conflicts = porcelain.merge_tree(
                                ^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4593, in merge_tree
    merged_tree, conflicts = merger.merge_trees(base, ours, theirs)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_merge_tree_with_tree_objects (tests.test_porcelain_merge.PorcelainMergeTreeTests.test_merge_tree_with_tree_objects)
Test merge_tree with tree objects instead of commits.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_porcelain_merge.py", line 444, in test_merge_tree_with_tree_objects
    merged_tree_id, conflicts = porcelain.merge_tree(
                                ^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/porcelain.py", line 4593, in merge_tree
    merged_tree, conflicts = merger.merge_trees(base, ours, theirs)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

======================================================================
ERROR: test_rebase_with_conflicts (tests.test_rebase.RebaserTestCase.test_rebase_with_conflicts)
Test rebase with merge conflicts.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mkatafiasz/Dev/dulwich/tests/test_rebase.py", line 211, in test_rebase_with_conflicts
    rebase(self.repo, b"refs/heads/master", branch=b"refs/heads/feature")
  File "/home/mkatafiasz/Dev/dulwich/dulwich/rebase.py", line 929, in rebase
    result = rebaser.continue_()
             ^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/rebase.py", line 803, in continue_
    new_sha, conflicts = self._cherry_pick(commit, onto)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/rebase.py", line 701, in _cherry_pick
    merged_tree, conflicts = three_way_merge(
                             ^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 503, in three_way_merge
    return merger.merge_trees(base_tree, ours_tree, theirs_tree)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 436, in merge_trees
    merged_content, had_conflict = self.merge_blobs(
                                   ^^^^^^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 267, in merge_blobs
    return merge_blobs(
           ^^^^^^^^^^^^
  File "/home/mkatafiasz/Dev/dulwich/dulwich/merge.py", line 216, in merge_blobs
    m = merge3.Merge3(
        ^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'Merge3'

----------------------------------------------------------------------
Ran 2846 tests in 194.443s

FAILED (errors=10, skipped=157, expected failures=1)
make: *** [Makefile:25: check] Error 1
```

All are seemingly related to the missing `merge3` import in `merge.py`. I haven't installed the optional dependencies, so it makes sense the import isn't there, but since the module very clearly intends to make that an optional import which isn't a failure if missing, bombing later on in use seems wrong. It should either be able to do its job even in the absence of `merge3`, or clearly indicate that requested functionality is not available without it.

--- a/NEWS
+++ b/NEWS
@@ -1,5 +1,9 @@
 0.24.2	UNRELEASED
 
+ * Fix merge functionality to gracefully handle missing optional merge3 dependency
+   by raising informative ImportError with installation instructions.
+   (Jelmer Vernooĳ, #1759)
+
  * Fix worktree CLI tests to properly change to repository directory.
    (Jelmer Vernooĳ, #1738)
 

--- a/dulwich/merge.py
+++ b/dulwich/merge.py
@@ -163,6 +163,11 @@ def merge_blobs(
             return ours_blob.data, False
         else:
             # Both added different content - conflict
+            if merge3 is None:
+                raise ImportError(
+                    "merge3 module is required for three-way merging. "
+                    "Install it with: pip install merge3"
+                )
             m = merge3.Merge3(
                 [],
                 ours_blob.data.splitlines(True),
@@ -185,6 +190,11 @@ def merge_blobs(
                 return b"", False  # They didn't modify, accept deletion
             else:
                 # Conflict: we deleted, they modified
+                if merge3 is None:
+                    raise ImportError(
+                        "merge3 module is required for three-way merging. "
+                        "Install it with: pip install merge3"
+                    )
                 m = merge3.Merge3(
                     base_content.splitlines(True),
                     [],
@@ -197,6 +207,11 @@ def merge_blobs(
                 return b"", False  # We didn't modify, accept deletion
             else:
                 # Conflict: they deleted, we modified
+                if merge3 is None:
+                    raise ImportError(
+                        "merge3 module is required for three-way merging. "
+                        "Install it with: pip install merge3"
+                    )
                 m = merge3.Merge3(
                     base_content.splitlines(True),
                     ours_content.splitlines(True),
@@ -213,6 +228,11 @@ def merge_blobs(
         return ours_content, False
 
     # Perform three-way merge
+    if merge3 is None:
+        raise ImportError(
+            "merge3 module is required for three-way merging. "
+            "Install it with: pip install merge3"
+        )
     m = merge3.Merge3(
         base_content.splitlines(True),
         ours_content.splitlines(True),

--- a/tests/test_merge_drivers.py
+++ b/tests/test_merge_drivers.py
@@ -20,6 +20,7 @@
 
 """Tests for merge driver support."""
 
+import importlib.util
 import sys
 import unittest
 from typing import Optional
@@ -209,6 +210,10 @@ class MergeBlobsWithDriversTests(unittest.TestCase):
 
     def setUp(self):
         """Set up test fixtures."""
+        # Check if merge3 module is available
+        if importlib.util.find_spec("merge3") is None:
+            raise unittest.SkipTest("merge3 module not available, skipping merge tests")
+
         # Reset global registry
         global _merge_driver_registry
         from dulwich import merge_drivers

